sql/src/main/java/org/apache/druid/sql/AbstractStatement.java

```diff
@@ -0,0 +1,203 @@
+
+  protected final SqlToolbox sqlToolbox;
+  protected final SqlQueryPlus queryPlus;
+  protected final SqlExecutionReporter reporter;
+  protected PlannerContext plannerContext;
+  protected Set<ResourceAction> queryResource;
+  protected Set<ResourceAction> resourceActions;
+
+  public AbstractStatement(
+      final SqlToolbox sqlToolbox,
+      final SqlQueryPlus sqlRequest,
+      final String remoteAddress
+  )
+  {
+    this.sqlToolbox = sqlToolbox;
+    this.queryPlus = sqlRequest;
+    this.reporter = new SqlExecutionReporter(this, remoteAddress);
+
+    // Context is modified, not copied.
+    contextWithSqlId(sqlRequest.context())
+      .addDefaultParams(sqlToolbox.defaultQueryConfig.getContext());
+  }
+
+  private QueryContext contextWithSqlId(QueryContext queryContext)
+  {
+    // "bySegment" results are never valid to use with SQL because the result format is incompatible
+    // so, overwrite any user specified context to avoid exceptions down the line
+
+    if (queryContext.removeUserParam(QueryContexts.BY_SEGMENT_KEY) != null) {
+      log.warn("'bySegment' results are not supported for SQL queries, ignoring query context parameter");
+    }
+    queryContext.addDefaultParam(PlannerContext.CTX_SQL_QUERY_ID, UUID.randomUUID().toString());
+    return queryContext;
+  }
+
+  public String sqlQueryId()
+  {
+    return queryPlus.context().getAsString(PlannerContext.CTX_SQL_QUERY_ID);
+  }
+
+  /**
+   * Validate SQL query and authorize against any datasources or views which
+   * will take part in the query. Must be called by the API methods, not
+   * directly.
+   */
+  protected void validate(DruidPlanner planner)
+  {
+    plannerContext = planner.getPlannerContext();
+    plannerContext.setAuthenticationResult(queryPlus.authResult());
+    plannerContext.setParameters(queryPlus.parameters());
+    try {
+      planner.validate();
+    }
+    // We can't collapse catch clauses since SqlPlanningException has
+    // type-sensitive constructors.
+    catch (SqlParseException e) {
+      throw new SqlPlanningException(e);
+    }
+    catch (ValidationException e) {
+      throw new SqlPlanningException(e);
+    }
+  }
+
+  /**
+   * Authorize the query using the authorizer provided, and an option to authorize
+   * context variables as well as query resources.
+   */
+  protected void authorize(
+      DruidPlanner planner,
+      Function<Set<ResourceAction>, Access> authorizer
+  )
+  {
+    boolean authorizeContextParams = sqlToolbox.authConfig.authorizeQueryContextParams();
+
+    // Authentication is done by the planner using the function provided
+    // here. The planner ensures that this step is done before planning.
+    Access authorizationResult = planner.authorize(authorizer, authorizeContextParams);
+    if (!authorizationResult.isAllowed()) {
+      throw new ForbiddenException(authorizationResult.toString());
+    }
+
+    queryResource = planner.resourceActions(false);
```

Comment:
Why do we need to pass an explicit `false` here, instead of `authorizeContextParams`? Seems sketchy, so there should be a comment explaining this.

I also suggest having the names be more aligned, since they're very similar things. So, instead of `resourceActions` and `queryResource`, something like `resourceActions` and `resourceActionsWithoutContext`.